model/lifecycle.go

```diff
@@ -930,11 +931,42 @@ func createTasksForBuild(creationInfo TaskCreationInfo) (task.Tasks, error) {
 	return tasks, nil
 }
 
+// addSingleHostTaskGroupDependencies adds dependencies to tasks in a single-host task group
+func addSingleHostTaskGroupDependencies(taskMap map[string]*task.Task, p *Project, taskIds TaskIdTable) {
+	for _, t := range taskMap {
+		if t.TaskGroup == "" {
+			continue
+		}
+		tg := p.FindTaskGroup(t.TaskGroup)
+		if tg == nil || tg.MaxHosts > 1 {
+			continue
+		}
+		singleHostTGDeps := []TaskUnitDependency{}
+		// Iterate backwards until we find a task that exists in the taskMap. This task
+		// will be the parent dependency for the current single host TG task.
+		taskFound := false
+		for i := len(tg.Tasks) - 1; i >= 0; i-- {
+			if t.DisplayName == tg.Tasks[i] {
+				taskFound = true
+				continue
+			}
+			if _, ok := taskMap[taskIds.GetId(t.BuildVariant, tg.Tasks[i])]; ok && taskFound {
```

Comment:
Is there ever a case where we could expect `ok` would be false here? I see there's [a test for this](https://github.com/evergreen-ci/evergreen/blob/1a8abd172c2f87f084c5f765e61cce7def610f05/model/lifecycle_test.go#L1500), but I'm not sure if/when this could happen.

Is it worth logging or something if `ok` is ever false?

Reply:
This could happen if a task group contains `t1, t2, t3` and we only create `t1` and `t3`. Since `t2` doesn't exist then `t3` needs to depend directly on `t1`. The check for `ok` is just to see if the task was actually created, and if not we can just skip and look for the next task in the group when we're adding dependencies.

Reply:
You can schedule a subset of  a single host task group tasks in the UI. If they succeed currently, there will be no dependencies ([example](https://spruce-staging.corp.mongodb.com/version/63bc8508b237360254cdef85/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)), but if one fails, dependencies will be set ([example](https://spruce-staging.corp.mongodb.com/version/63bc853397b1d313d5bf02c3/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)).

Reply:
Interesting... it seems a bit strange that it's allowed, but I guess it doesn't matter much as long as the dependencies are set up correctly.